### PR TITLE
provider/azure: retry requests on status code 429

### PR DESCRIPTION
**Pull request by axw**
If Azure returns status code 429 (TooManyRequests), then that means we're
being rate-limited and we should back off and retry later. The 429
responses do not contain a Retry-After field, so we just use exponential
backoff starting at 5 seconds, going up to 1 minute between retries, with a
maximum of 5 minutes per API request.

Also updated Azure to the latest SDK.

Fixes https://bugs.launchpad.net/juju-core/+bug/1540394

(Review request: http://reviews.vapour.ws/r/4816/)

### CODE REVIEW

File: provider/azure/environ.go [hunk 1]
```diff
@@ -176,15 +180,19 @@ func createStorageAccount(
...
-			storage.AccountCheckNameAvailabilityParameters{
-				Name: to.StringPtr(accountName),
-				// Azure is a little inconsistent with when Type is
-				// required. It's required here.
-				Type: to.StringPtr("Microsoft.Storage/storageAccounts"),
-			},
-		)
-		if err != nil {
+		var result storage.CheckNameAvailabilityResult
+		if err := callAPI(func() (autorest.Response, error) {
```
Comment (wallyworld) [inline provider/azure/environ.go]: Should we define a type for this func signature

File: provider/azure/environ.go [hunk 2]
```diff
@@ -836,14 +869,18 @@ func deleteInstance(
 	// Delete the VM's OS disk VHD.
 	logger.Debugf("- deleting OS VHD")
 	blobClient := storageClient.GetBlobService()
-	if _, err := blobClient.DeleteBlobIfExists(osDiskVHDContainer, vmName); err != nil {
+	if err := callAPI(func() (autorest.Response, error) {
+		_, err := blobClient.DeleteBlobIfExists(osDiskVHDContainer, vmName)
```
Comment (wallyworld) [inline provider/azure/environ.go]: Curious - why is this api not applicable to rate limiting?

**Comment by wallyworld** [comment]
LGTM

File: provider/azure/environ.go (hunk 1 above)
Comment (axw) [inline provider/azure/environ.go]: Wouldn't help. You can't use a named function type in a function literal like that.

File: provider/azure/environ.go (hunk 2 above)
Comment (axw) [inline provider/azure/environ.go]: It's a different Azure API altogether. Actually there's no point that being wrapped in callAPI, I'll take it back out.

**Comment by wallyworld** [comment]
$$fixes-1540394$$

**Pull request merged**